Don't record satisfied wall clock triggers in the DB

wxtim opened the pull request:
Closes https://github.com/cylc/cylc-flow/issues/5911

n.b. I tried to find a more sophisticated method than `startswith('wall_clock(...')` to identify the correct xtriggers, but I think this will do. 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] This is a change to internals, so it doesn't need a docs/changelog update.
- [x] Raised against 8.2.x

oliver-sanders commented:
Execution/submission retries are also implemented via wall_clock xtriggers.

Worth checking/testing that these aren't written to the DB.

wxtim commented on tests/integration/test_workflow_db_mgr.py:
```diff
@@ -127,3 +128,53 @@ async def test_workflow_param_rapid_toggle(
+            'foo': {
+                'execution retry delays': 'PT0S',
+                'script': (
+                    'test $CYLC_TASK_SUBMIT_NUMBER == 1 && exit 1 || exit 0')
```
This will be do-able in simulation mode after the simulation mode refactoring, which should speed this test up rather a lot.

oliver-sanders replied:
Shell interaction is the line in the sand between integration and functional tests.

But I expect it can be dispensed with, here's one idea, get the simulated task to wait so you can feed the failed message into the system manually:

```ini
[runtime]
  [[foo]]
     [[[simulation]]]
       default run length = PT1M
```

```python
# fail a simulated task
schd.task_events_mgr.process_message(itask, 'error', 'failed')
```

wxtim replied:
@oliver-sanders  - Using simulation mode isn't possible because of the bug outlined in https://github.com/cylc/cylc-flow/issues/5935.

wxtim commented on tests/integration/test_workflow_db_mgr.py:
```diff
@@ -127,3 +128,53 @@ async def test_workflow_param_rapid_toggle(
+    schd = scheduler(id_, paused_start=False, run_mode='live')
+    async with run(schd) as log:
+        while 'Workflow shutting down - AUTOMATIC' not in log.messages:
+            await asyncio.sleep(0.5)
```
This seems to be the best balance between waiting longer and checking more - but it won't get < 15s.

oliver-sanders replied:
The `complete` fixture can do this for you. I think it's:

```python
await complete(schd)
```

wxtim replied:
There's a complete_task fixture?

MetRonnie commented on tests/integration/test_workflow_db_mgr.py:
```diff
@@ -127,3 +128,53 @@ async def test_workflow_param_rapid_toggle(
+        while 'Workflow shutting down - AUTOMATIC' not in log.messages:
+            await asyncio.sleep(0.5)
+
+    # Get xtriggers db table:
+    info = schd.workflow_db_mgr.get_pri_dao().conn.execute(
+        'SELECT * FROM xtriggers').fetchall()
+
+    # All xtriggers are xrandom: None are wall_clock:
+    assert all(i[0].startswith('xrandom') for i in info)
```
Can simply use `db_select` fixture
```suggestion
    assert db_select(schd, False, 'xtriggers', 'signature') == [
        ('xrandom(100)',),
        ('xrandom(100, _=Not a real wall clock trigger)',)
    ]
```

oliver-sanders commented on cylc/flow/workflow_db_mgr.py:
```diff
@@ -405,9 +405,10 @@ def put_task_event_timers(self, task_events_mgr):
-            self.db_inserts_map[self.TABLE_XTRIGGERS].append({
-                "signature": sig,
-                "results": json.dumps(res)})
+            if not sig.startswith('wall_clock(trigger_time='):
```
Probably safer not to check the kwarg incase it's provided by position:

```suggestion
            if not sig.startswith('wall_clock('):
```

oliver-sanders commented on tests/integration/conftest.py:
```diff
@@ -473,3 +475,85 @@ def _inner(source, **kwargs):
+
+
+@pytest.fixture
+def complete():
```
Unused.

MetRonnie replied:
It was supposed to be used as per https://github.com/cylc/cylc-flow/pull/5923#discussion_r1453217855

oliver-sanders replied:
The test for completion was changed since then, the new test is reasonable providing it is constrained by a timeout. Otherwise switch to `complete()` which has a default timeout built in.

oliver-sanders commented on tests/integration/test_workflow_db_mgr.py:
```diff
@@ -127,3 +132,54 @@ async def test_workflow_param_rapid_toggle(
+        foo = schd.pool.get_task(ISO8601Point(rawpoint), 'foo')
+
+        # Wait until all xtriggers for foo are satisfied:
+        while not all(foo.state.xtriggers.values()):
+            await asyncio.sleep(1)
```
Should wrap this in `async_timeout` to avoid this test hanging indefinitely if something goes wrong.

Try `git grep async_tmeout` for example usage.

MetRonnie reviewed (changes requested):
`complete` fixture ported from master but still needs to be used in the test

MetRonnie reviewed (approved):
Not merging pending async timeout comment above

wxtim commented:
@oliver-sanders  - quick check that https://github.com/cylc/cylc-flow/pull/5923/commits/94f3a769360bc78c639f45d9ec25442244c73e5e addresses your comments?
